File: 1669-minimum-cost-to-cut-a-stick/1669-minimum-cost-to-cut-a-stick.py
class Solution:
    def minCost(self, n: int, cuts: List[int]) -> int:
        cuts.sort()
        memo = {}
        def dp(start, length):
            if (start, length) in memo: return memo[(start, length)]
            ans = inf

            for cut in cuts[bisect_right(cuts, start):]:
                if cut >= start + length: break

                leftLen = cut - start
                rightLen = length - leftLen

                ans = min(ans, length + dp(start, leftLen) + dp(cut, rightLen))
            memo[(start, length)] = ans if ans != inf else 0
            return ans if ans != inf else 0
        
        return dp(0, n)

File: 1669-minimum-cost-to-cut-a-stick/1669-minimum-cost-to-cut-a-stick.cpp
class Solution {
public:
    int minCost(int n, vector<int>& cuts) {
        sort(cuts.begin(), cuts.end());
        vector<int> extendedCuts;
        extendedCuts.push_back(0);
        for(int i = 0; i < cuts.size(); i++) {
            extendedCuts.push_back(cuts[i]);
        }
        extendedCuts.push_back(n);

        vector<vector<long>> memo(extendedCuts.size(), vector<long>(extendedCuts.size(), -1));
        function<long(int, int)> dp = [&] (int start, int end) {
            if (start + 1 == end) {
                return 0l;
            }

            if (memo[start][end] != -1) return memo[start][end];

            long ans = INT_MAX;
            for (int i = start + 1; i < end; i++) {
                ans = min(ans, + extendedCuts[end] - extendedCuts[start] + dp(start, i) + dp(i, end));
            }

            memo[start][end] = ans;
            return ans;
        };

        return dp(0, extendedCuts.size() - 1);
    }
};